src/goto-programs/remove_const_function_pointers.cpp

```diff
@@ -0,0 +1,1013 @@
+}
+
+/*******************************************************************\
+
+Function: remove_const_function_pointerst::try_resolve_typecast_function_call
+
+  Inputs:
+   typecast_expr - The typecast expression to resolve to possible function calls
+   out_functions - The functions this expression could be
+
+ Outputs: Returns true if it was able to resolve the typecast expression to
+          some specific functions. If this is the case, out_functions will
+          contain the possible functions.
+
+ Purpose: To resolve an expression to the specific function calls it can
+          be. Specifically, this function deals with typecast expressions
+          by looking at the type cast values.
+
+\*******************************************************************/
+
+bool remove_const_function_pointerst::try_resolve_typecast_function_call(
+  const typecast_exprt &typecast_expr, functionst &out_functions)
+{
+  // We simply ignore typecasts and assume they are valid
+  // I thought simplify_expr would deal with this, but for example
+  // a cast from a 32 bit width int to a 64bit width int it doesn't seem
+  // to allow
+  functionst typecast_values;
+  bool resolved=
+    try_resolve_function_call(typecast_expr.op(), typecast_values);
+
+  if(resolved)
+  {
+    out_functions.insert(typecast_values.begin(), typecast_values.end());
+    return true;
+  }
+  else
+  {
+    LOG("Failed to squash typecast", typecast_expr);
+    return false;
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_const_function_pointerst::try_resolve_expression
+
+  Inputs:
+   expr - The expression to try and squash
+   out_resolved_expression - The squashed version of this expression
+   out_is_const - Is the squashed expression constant
+
+ Outputs: Returns true providing the squashing went OK (note it
+          may not have squashed anything). The out_resolved_expression will in
+          this case be all the possible squashed versions of the supplied
+          expression.
+          The out_is_const will return whether the squashed value is suitably
+          const (e.g. if we squashed a struct access, was the struct const).
+
+ Purpose: To squash various expr types to simplify the expression.
+          ID_index -> dig to find ID_array and get the values out of it
+          ID_member -> dig to find ID_struct and extract the component value
+          ID_dereference -> dig to find ID_address_of and extract the value
+          ID_typecast -> return the value
+          ID_symbol -> return false, const symbols are squashed first and
+                       non const symbols cannot be squashed
+          Everything else -> unchanged
+
+\*******************************************************************/
+
+bool remove_const_function_pointerst::try_resolve_expression(
+  const exprt &expr, expressionst &out_resolved_expression, bool &out_is_const)
+{
+  const exprt &simplified_expr=simplify_expr(expr, ns);
```

Comment:
`simplify_expr` returns by value, thus you don't really want the reference here.